flash/core/model.py

```diff
@@ -143,48 +154,137 @@ def predict(
+            return self.datamodule.data_pipeline
+
+        if self.trainer is not None and hasattr(
+            self.trainer, 'datamodule'
+        ) and getattr(self.trainer.datamodule, 'data_pipeline', None) is not None:
+            return self.trainer.datamodule.data_pipeline
         return self._data_pipeline
 
     @data_pipeline.setter
     def data_pipeline(self, data_pipeline: DataPipeline) -> None:
         self._data_pipeline = data_pipeline
+        if data_pipeline is not None and getattr(data_pipeline, '_preprocess_pipeline', None) is not None:
+            self._preprocess = data_pipeline._preprocess_pipeline
+
+        if data_pipeline is not None and getattr(data_pipeline, '_postprocess_pipeline', None) is not None:
+            datapipeline_postprocess = getattr(data_pipeline, '_postprocess_pipeline', None)
+            if type(datapipeline_postprocess) != Postprocess:
+                self._postprocess = data_pipeline._postprocess_pipeline
+
+    def on_train_start(self) -> None:
+        if self.data_pipeline is not None:
+            self.data_pipeline._attach_to_model(self, RunningStage.TRAINING)
+        return super().on_train_start()
+
+    def on_train_end(self) -> None:
+        if self.data_pipeline is not None:
+            self.data_pipeline._detach_from_model(self)
+        return super().on_train_end()
+
+    def on_validation_start(self) -> None:
+        if self.data_pipeline is not None:
+            self.data_pipeline._attach_to_model(self, RunningStage.VALIDATING)
+        return super().on_validation_start()
+
+    def on_validation_end(self) -> None:
+        if self.data_pipeline is not None:
+            self.data_pipeline._detach_from_model(self)
+        return super().on_validation_end()
+
+    def on_test_start(self) -> None:
+        if self.data_pipeline is not None:
+            self.data_pipeline._attach_preprocess_to_model(self, RunningStage.TESTING)
+        return super().on_test_start()
+
+    def on_test_end(self):
+        if self.data_pipeline is not None:
+            self.data_pipeline._detach_from_model(self)
+        return super().on_test_end()
+
+    def on_predict_start(self):
+        if self.data_pipeline is not None:
+            self.data_pipeline._attach_to_model(self, RunningStage.PREDICTING)
+
+        return super().on_predict_start()
+
+    def on_predict_end(self):
+        if self.data_pipeline is not None:
+            self.data_pipeline._detach_from_model(self)
+        return super().on_predict_end()
 
-    @staticmethod
-    def default_pipeline() -> DataPipeline:
-        """Pipeline to use when there is no datamodule or it has not defined its pipeline"""
-        return DataModule.default_pipeline()
+    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        # TODO: Is this the best way to do this? or should we also use some kind of hparams here?
```

Comment:
@tchaton @carmocca Thoughts on this? Even though PyTorch does not recommend this, I think we should go with this approach which just pickles the entire class, since this is also the one we want to use for prediction lateron.